app/org/zalando/hutmann/authentication/OAuth2Action.scala

```diff
@@ -129,39 +135,63 @@ class OAuth2Action(
   /**
     * Add the authorization information to the request.
     */
-  def transform[A](request: Request[A]): Future[UserRequest[A]] = {
-    implicit val context: RequestContext = request
-    val tokenOpt: Option[String] = getToken(request)
+  def authenticate(requestHeader: RequestHeader): Future[Either[AuthorizationProblem, User]] = {
+    implicit val context: RequestContext = requestHeader
+    val tokenOpt: Option[String] = getToken(requestHeader)
     tokenOpt match {
       case Some(token) =>
         val futureToken = validateToken(token).recoverWith {
           case NonFatal(ex) =>
             logger.warn("Problem getting OAuth token, retrying...", ex)
             validateToken(token)
-        }.flatMap{
+        }.flatMap {
           case Left(TimeoutAuthError) =>
             logger.warn("Gateway timeout while getting OAuth token, retrying...")
             validateToken(token)
           case other => Future.successful(other)
         }
-        val futureUser = futureToken.flatMap(tokenResultScoring).recoverWith({
+        futureToken.flatMap(tokenResultScoring).recoverWith({
           case _: TimeoutException => Future.successful(Left(AuthorizationTimeout))
           case e                   => throw e
         })
-        futureUser.map(new UserRequest(_, request))
       case None =>
         logger.info("No authorization founder in 'Authorization' header or 'access_token' query parameter. Rejecting access.")
-        Future.successful(new UserRequest(Left(NoAuthorization), request))
+        Future.successful(Left(NoAuthorization))
     }
   }
 
   override def invokeBlock[A](request: Request[A], block: (UserRequest[A]) => Future[Result]): Future[Result] = {
-    transform(request).flatMap(userRequest =>
+    authenticate(request).flatMap { user =>
+      implicit val context: RequestContext = request
+      val userRequest = new UserRequest[A](user, request)
       if (autoReject) {
         autoRejectBehaviour(userRequest.user, block(userRequest))
       } else {
         block(userRequest)
-      }) recoverWith recoveryBehaviour(request)
+      }
+    } recoverWith recoveryBehaviour(request)
+  }
+
+  def essentialAction[A](bodyParser: BodyParser[A])(block: UserRequest[A] => Future[Result]): EssentialAction = {
+    implicit val actorSystem = ActorSystem("hutmann-actor-system")
+    implicit val materializer = ActorMaterializer(namePrefix = Some("hutmann-actor-materializer"))
```

Comment:
@hardvain Expected `materializer` as implicit parameter in apply method, and removed creation of `ActorSystem`.